Create PSA configs for ECC curves

d3zd3z opened the pull request:
PSA configuration defines for ECC curves. This implements #3678. There is a 1:1 correspondence between the traditional configuration defines and the PSA_WANT defines.

## Status
**READY**

## Requires Backporting
No backporting (PSA defines)

d3zd3z commented:
Just the ifdef part, but I wanted to push early to get any feedback if I didn't get the names right.

gilles-peskine-arm commented on include/mbedtls/config_psa.h:
```diff
@@ -355,6 +355,90 @@ extern "C" {
 #endif /* PSA_WANT_KEY_TYPE_CHACHA20 */
 #endif /* PSA_WANT_ALG_CHACHA20_POLY1305 */
 
+#if defined(PSA_WANT_ECC_SECP192R1)
```
```suggestion
#if defined(PSA_WANT_ECC_SECP_R1_192)
```
See https://github.com/ARMmbed/mbedtls/blob/2c5d9e6a32ec2e131eaf1c015dab65b2ff83059e/docs/proposed/psa-conditional-inclusion-c.md (the document is already in the tree but https://github.com/ARMmbed/mbedtls/pull/4133 clarifies that section).

The macro names have this structure so that they can be derived mechanically from the `PSA_ECC_FAMILY_xxx` symbol and the key size.

d3zd3z commented:
I've updated the names according to the proposed psa conditional inclusion doc. I can squash these if that seems appropriate, rather than adding a whole commit that adds incorrect names and another that fixes it.

stevew817 commented on include/psa/crypto_config.h:
```diff
@@ -80,6 +80,20 @@
 #define PSA_WANT_ALG_TLS12_PSK_TO_MS            1
 #define PSA_WANT_ALG_XTS                        1
 
+#define PSA_WANT_ECC_BP256R1                    1
```
Looks like you forgot to update these names, @d3zd3z

d3zd3z replied:
@stevew817 thanks, fixed.

stevew817 commented:
Travis doesn't agree...

stevew817 commented:
@d3zd3z Looks like there's still something in the generated cases that doesn't quite play ball

d3zd3z commented:
I removed the revert, as the workaround in d169d6031975640ecc96053daf5f1fb9c19968f1 is still needed for unimplemented config options. Instead, I regenerated the list, adding back in the ones I've implemented. Also fixed the typo for the one that was actually meaningfully failing.

Looks like I need to rebase, though.

mpg commented:
@d3zd3z  In the pr-merge job, the failing tests are `all.sh` components `test_depends_pkalgs` and `test_depends_pkalgs_psa` so this might be the same root cause for both. In the pr-head job, the same components failed, and in addition one of the windows builds failed to, with a cryptic linker error, while all other windows builds passed, so I'm inclined to disregard it for now and see if it passes in the next iteration.

d3zd3z commented:
reviewers: I found the test failure was caused by an incorrect resolution during a rebase. I've added a new patch, but if you'd prefer, I can fold that into the original commit to correct the rebase. Let me know.

ronald-cron-arm commented on include/psa/crypto_config.h:
```diff
@@ -80,6 +80,20 @@
 #define PSA_WANT_ALG_TLS12_PSK_TO_MS            1
 #define PSA_WANT_ALG_XTS                        1
 
+#define PSA_WANT_ECC_BP_R1_256                  1
```
I guess this is for curves belonging to the Brainpool P family. Thus to align with the PSA spec this should rather be:
PSA_WANT_ECC_BRAINPOOL_P_R1_256. Please change wherever applicable.

ronald-cron-arm commented on include/mbedtls/config_psa.h:
```diff
@@ -355,6 +355,97 @@ extern "C" {
 #endif /* PSA_WANT_KEY_TYPE_CHACHA20 */
 #endif /* PSA_WANT_ALG_CHACHA20_POLY1305 */
 
+#if defined(PSA_WANT_ECC_SECP_R1_192)
```
Please re-order in alphabetic order.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -138,95 +138,95 @@ depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:MBEDTLS_RSA_C
 
 PSA import/export EC secp224r1 key pair: good
-depends_on:PSA_WANT_ALG_ECDSA:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:MBEDTLS_ECP_DP_SECP224R1_ENABLED
+depends_on:PSA_WANT_ALG_ECDSA:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:PSA_WANT_ECC_SECP_R1_224
```
I guess you've done those changes by modifying set_psa_test_dependencies.py and then running it. Please commit the changes in set_psa_test_dependencies.py as well.

d3zd3z replied:
I did not have to change this script to get these values. It looks like it is extracting the PSA_WANT values itself. There are mentions of the old curve names, but in comments. I can fix that if it would be useful.

ronald-cron-arm replied:
I forgot that set_psa_test_dependencies.py needs updates not related to this PR (mostly due to the merge of #4009) thus let's forget about it for this PR.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_not_supported.generated.data:
```diff
@@ -372,138 +372,138 @@ depends_on:PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY:!PSA_WANT_ECC_BRAINPOOL_P_R1_512:DEP
 
 PSA import ECC_KEY_PAIR(MONTGOMERY) 255-bit type not supported
-depends_on:!PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_MONTGOMERY_255:DEPENDENCY_NOT_IMPLEMENTED_YET
+depends_on:!PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_MONTGOMERY_255
```
Please commit the changes to the Python scripts used to do those changes.

d3zd3z replied:
As far as I can tell, the script updated these based on these defines being added to the config_psa.h. I did not modify the python scripts.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_se_driver_hal.data:
```diff
@@ -163,25 +163,25 @@ Key registration: key id max volatile
 
 Import-sign-verify: sign in driver, ECDSA
-depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED
+depends_on:MBEDTLS_ECDSA_C:MBEDTLS_ECP_C:PSA_WANT_ECC_SECP_R1_256
```
I am not sure why in fact but in the previous similar cases we didn't update the dependencies of test_suite_psa_crypto_se_driver_hal*.data files (see 8405b5f6). @gilles-peskine-arm why this exception ?

gilles-peskine-arm replied:
I thought that the mock SE drivers in these tests might not work with a new-style accelerator driver. As long as the CI passes, and will keep passing when we improve coverage with accelerator drivers, it doesn't matter. Old-style SE support is currently on suspended animation: we don't try to improve it, but we do try not to break what's already there.

ronald-cron-arm replied:
ok thanks, Gilles. Thus I guess better to stay on the same path and do not update the dependencies in test_suite_psa_crypto_se_driver_hal.data with PSA_WANT* macros.

ronald-cron-arm commented on include/psa/crypto_config.h:
```diff
@@ -80,9 +80,9 @@
-#define PSA_WANT_ECC_BP_R1_256                  1
-#define PSA_WANT_ECC_BP_R1_384                  1
-#define PSA_WANT_ECC_BP_R1_512                  1
+#define PSA_WANT_ECC_BRAINPOOL_P_R1_256                  1
```
Please align.

ronald-cron-arm reviewed (changes requested):
Just an alignment issue and this [pending comment](https://github.com/ARMmbed/mbedtls/pull/4158#discussion_r594404559). @gilles-peskine-arm could you have a look to the comment ?

ronald-cron-arm commented:
Thanks in fixing the alignment. There is only [this pending comment remaining](https://github.com/ARMmbed/mbedtls/pull/4158#discussion_r594404559): please revert the changes to test_suite_psa_crypto_se_driver_hal.data where we prefer to stick with dependencies based on MBEDTLS_ macros. I've checked with the full configuration that the test suite is running successfully with that change thus it should be ok with the CI.

gilles-peskine-arm commented:
> please revert the changes to test_suite_psa_crypto_se_driver_hal.data where we prefer to stick with dependencies based on MBEDTLS_ macros.

I wouldn't say that we prefer to stick with old dependencies: if the new dependencies work, I'm fine with that. My intent with this file was to minimize work. Since the curve dependencies have been updated, let's keep the new ones.

ronald-cron-arm commented:
ok fine by me. Approving then.

gilles-peskine-arm commented:
I'm afraid I can't fit a review in my current schedule.

ronald-cron-arm commented:
That's ok @gilles-peskine-arm I just want to know if we need to find another reviewer or not. @stevew817 what about you?

stevew817 reviewed (approved):
LGTM
